### ISSUE TITLE
If you start drag but never drop with mouse, this will kill i3lock!!

### ISSUE THREAD

**Issue by artheus** [i3/i3lock#98]
I consider this to be a security issue of the i3lock.

If someone would connect a secondary mouse to my computer at work. And whenever I lock my computer they will click the mouse button and not release it, this will kill i3lock after a little while.

So steps to reproduce this:

1. Click mouse button (start drag)
2. Start i3lock
3. w/o releasing the mouse button, wait for a little while
Boom! i3lock dies!

debug output is:
```
[i3lock-debug] device = 3
[i3lock-debug] found Xinerama screen: 1920 x 1080 at 0 x 0
[i3lock-debug] found Xinerama screen: 1920 x 1200 at 1920 x 0
[i3lock-debug] scaling_factor is 1, physical diameter is 190 px
[i3lock-debug] Watching window 0x01e00003
i3lock: Cannot grab pointer/keyboard
[i3lock-debug] Read event of type 18
[i3lock-debug] UnmapNotify for 0x01e00003
```

**Comment by stapelberg** [comment]
This should be fixed with commit https://github.com/i3/i3lock/commit/1c97a8484723b7db6c96c3b7bfa4c86c26f47b5a already. Could you confirm please?

**Comment by artheus** [comment]
I cloned the latest master, and build it locally.
```
$ i3lock -v
i3lock: version 2.8-7-g1c97a84 (2016-09-27) © 2010 Michael Stapelberg
```

This is the version I got from doing that.
I can not really understand if the `g1c97a84` part should be pointing to the latest commit. But from doing a git log I can see that the fix of issue #88 is the latest commit.

This do not seem to fix the problem though. Following the exact same steps from my first comment  in this issue results in the same issue, killing i3lock.

**Comment by artheus** [comment]
My setup is Fedora
```
$ uname -a
Linux HOSTNAME 4.7.2-201.fc24.x86_64 #1 SMP Fri Aug 26 15:58:40 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```
(only replaced the hostname)

I use i3 as window manager (ofcourse)
```
$ i3 --version
i3 version 4.12 (2016-03-06, branch "4.12") © 2009 Michael Stapelberg and contributors
```

**Comment by Airblader** [comment]
It's not supposed to stop i3lock from preventing. We simply cannot force locking because X doesn't allow us to lock, this i3lock gives up. The referenced commit is simply supposed to give you feedback on whether the screen is already locked or still in the process of attempting it.

**Comment by artheus** [comment]
The only difference I see, now, is that I get a blue circle in the middle of the screen saying "Locking..".
@Airblader the feedback do not seem to be sufficient for actually telling me what is happening.

**Comment by Airblader** [comment]
@artheus What feedback would you expect?

**Comment by artheus** [comment]
@Airblader I would expect something like "Unable to lock screen, can not lock pointer/keyboard" or something like that. Maybe a shorter timeout for that feedback, and if it is finally able to lock, say something like "False alarm, everything is fine now"

Not necessarily exactly that of course, but something that will tell me more what is going on. Also give me the time to read it.

**Comment by Airblader** [comment]
I think that your approach is just as fine and logical, but not inherently better than what we have and thus I don't see a reason to change it. »Locking…« signals clearly (IMHO) that the locking process is still in progress and not yet finished. So the user should wait for it to finish. Once we lock the screen, there's no further need for a message.

> Maybe a shorter timeout for that feedback

Decreasing the time we keep trying increases the failure rate, which doesn't seem like a good trade-off.

**Comment by artheus** [comment]
@Airblader I think you might have misinterpreted what I meant by decreasing the timeout.

Right now the `Locking` text is displayed less than a second before the whole thing fails. So what I meant is decreasing the timeout before showing the text. I did not explain that properly.

**Comment by Airblader** [comment]
> Right now the Locking text is displayed less than a second before the whole thing fails.

What happens (just as an FYI) is that we don't display the message until some time has passed. The rationale for this decision was that in 99.9% of the cases, locking the screen will work just fine and displaying the message right away would lead to an ugly flash of the message.

However, I agree with you that it isn't too helpful if the message only shows up for a split second in the end either. Perhaps a better approach would be: try grabbing once; if it works, great; if it doesn't, start showing the message. Thoughts? @stapelberg @eplanet

**Comment by stapelberg** [comment]
We currently show the message after a certain time instead of after the first attempt because the first attempt often fails (very common case: people use `bindsym` to trigger i3lock instead of `bindsym --release`).

How about displaying an error message for 0.5s? E.g. “locking failed; exiting”

**Comment by artheus** [comment]
Ok. How many attempts are usually done before failing? Maybe show after
second fail?

I will look at the code to try and understand the case better.

**Comment by eplanet** [comment]
i3lock actually makes [10000 tries](https://github.com/i3/i3lock/blob/master/xcb.c#L173) before giving up. I personally think we should display the locking message quicker, but it doesn't make such a great difference. In any case, we cannot assume the time a given amount of tries will take.

Displaying an error message for 0.5s after failure is indeed a great idea that should please everybody, even though I'd rather prefer the lock message to be displayed a little longer, say for 2s, but that's only my opinion :smile:.

**Comment by Airblader** [comment]
> i3lock actually makes 10000 tries before giving up.

Yeah, because that's what the XCB documentation recommended (I'm pretty sure it was in the man page, but can't find it anymore – perhaps it was removed.) Other programs do the same.

> very common case: people use bindsym to trigger i3lock instead of bindsym --release

Given that displaying a message isn't really hurting the user as much as it is informing them of a misconfiguration I see that more as an argument to inform the user of it. We shouldn't encourage wasting tries on something the user can control that easily.

**Comment by eplanet** [comment]
Do you think this issue could be closed since #99 has been merged?

**Comment by Airblader** [comment]
I think it can, thanks!
